Design/Fit functions in python.md

```diff
@@ -0,0 +1,199 @@
+  c[1][0]['A0'] = 2
+  c['f1.f0.A0'] = 2
+```
+They will also define attributes with names equal to the names of the member functions + a suffix. The suffix is an int numbering repeating names. For example, members of function
+```
+  spectrum = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+```
+can be refered to like this:
+```
+  spectrum.LinearBackground
+  spectrum.Gaussian0
+  spectrum.Gaussian1
+```
+
+The wrappers will override `__iadd__(self, func)` and `__delitem__(self, i)` (`i` is an `int`, `func` is a `FitFunctionWrapper`) to implement adding and deleting members via `+=` and `del` operators:
+```
+  spectrum += Lorentzian(PeakCentre=3)
+  del spectrum[0]
+```
+
+Implement `__len__(self)` to return the number of member functions.
+```
+  n_peaks = len(spectrum)
+```
+
+Implement `__iter__` method to allow iteration over the members:
+```
+  for func in comp: 
+    print(func)
+```
+
+### Setting ties
+
+Ties are set with the `tie` method. Both keyword arguments and dictionaries of name/value pairs can be used as arguments.
+```
+  func.tie(A0=2.0)
+  func.tie({'f1.A2': '2*f0.A1', 'f2.A2': '3*f0.A1 + 1'})
+```
+
+Tying to a constant value that a parameter currently has can also be done with the `fix` method.
+```
+  func.fix('A0')
+  func.fix('f1.A2', 'f2.A2')
+```
+
+Ties are removed with the `untie` or `unfix` methods.
+```
+  func.unfix('f1.Sigma', 'f3.Sigma')
+  func.untie('f1.Sigma', 'f3.Sigma')
+```
+
+Composite functions should be able to set ties on all members with a single call:
+```
+  c.tieAll('Sigma')
+```
+The above call should tie all `Sigma` parameters in the member functions. It is equivalent to calling
+```
+  c.tie({'f1.Sigma': 'f0.Sigma', 'f2.Sigma': 'f0.Sigma', 'f3.Sigma': 'f0.Sigma', ...})
+```
+Similarly a call to `fixAll` fixes all parameters with the given name to their current value.
+```
+  spectrum.fixAll('FWHM')
+  spectrum.fixAll('f1.FWHM': 0.01)
+```
+Calling `fixAll()`/`unfixAll()` without arguments fixes/frees all parameters of a function.
+
+### Setting constraints
+
+Constraints are set with `constraints` method passing constraints as strings:
+```
+  g.constraints('-1 < PeakCentre < 1', 'Sigma > 0')
+```
+
+For composite functions constraints can be set both through the composite function and its members:
+```
+  c = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+  c.constraints('f1.Sigma > 0')
+  c[2].constraints('0 < Sigma < 1')
+```
+
+Composite functions should be able to set constraints on all members with a single call:
+```
+  c.constrainAll('Height > 1', 'Sigma > 0')
+```
+This should constrain the named parameters in all members that have them. Members that don't have these parameters should be skipped.
+
+## Fitting
+
+Python function `Fit` should accept instances of `FitFunctionWrapper` as its `Function` argument. `Fit` doesn't modify the input function but the returned value has the output function with the fitted parameters.
+
```

Comment:
I think we still need to process the output here: `out.Function` is a `FunctionProperty`, `out.Function.value` is an `IFunction`, but we need a `FitFunctionWrapper` or the user should wrap it.